File: src/09.tree/21.103.go
package tree

import "slices"

// BFS
func ZigzagLevelOrder(root *TreeNode) [][]int {
	ans := [][]int{}
	if root == nil {
		return ans
	}
	queue := []*TreeNode{root}
	k := 0

	for len(queue) > 0 {
		size := len(queue)
		temp := []int{}

		for i := size; i > 0; i-- {
			node := queue[0]
			queue = queue[1:]

			temp = append(temp, node.Val)

			if node.Left != nil {
				queue = append(queue, node.Left)
			}

			if node.Right != nil {
				queue = append(queue, node.Right)
			}
		}

		if k%2 == 1 {
			slices.Reverse(temp)
		}
		k++
		ans = append(ans, temp)
	}

	return ans
}

File: src/09.tree/21.103.ts
import { TreeNode } from "./01.100";

// BFS
export function zigzagLevelOrder(root: TreeNode | null): number[][] {
    const ans: number[][] = [];

    if (!root) {
        return ans;
    }

    const queue: TreeNode[] = [root];
    let k = 0;
    while (queue.length) {
        const size = queue.length;
        const temp: number[] = [];
        for (let i = size; i > 0; i--) {
            const node = queue.shift();
            if (k % 2 === 0) {
                temp.push(node.val);
            } else {
                temp.unshift(node.val);
            }
            node.left && queue.push(node.left);
            node.right && queue.push(node.right);
        }
        k++;
        ans.push(temp);
    }

    return ans;
}
